File: exercises/exercises-mpi-c/prime.c
/****************************************************************
 ****
 **** This program file is part of the book 
 **** `Parallel programming for Science and Engineering'
 **** by Victor Eijkhout, eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2012-8
 ****
 **** MPI Exercise for prime number factorization. So to speak.
 ****
 ****************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

int main(int argc,char **argv) {
  
  MPI_Init(&argc,&argv);
  MPI_Comm comm = MPI_COMM_WORLD;

  int nprocs, procno;
  MPI_Comm_size(comm,&nprocs);
  MPI_Comm_rank(comm,&procno);

  int bignum = 2000000111, maxfactor = 45200;
  // Exercise:
  // -- Parallelize the do loop so that each processor
  //    tries different candidate numbers.
  // -- If a processors finds a factor, print it to the screen.

  // 1. Set loop bounds
  int blocksize = maxfactor/nprocs;
  int myfirst = procno*blocksize,
    mylast = (procno+1)*blocksize-1 ;
  if (procno==0) myfirst = 2;
  if (procno==nprocs-1) mylast = maxfactor;
  // 2. Fill in loop header
  for ( int myfactor=
	  // this is sequential: 2; myfactor<bignum;
	  myfirst; myfactor<mylast;
	myfactor++
       ) {
    if (bignum%myfactor==0) {
      printf("Processor %d found factor %d\n",procno,myfactor);
    }
  }
  
  MPI_Finalize();
  return 0;
}

File: exercises/exercises-mpi-p/prime.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-8
####
#### prime.py : prime number finding
####
################################################################
################################################################

from mpi4py import MPI

comm = MPI.COMM_WORLD

bignum = 2000000111
maxfactor = 45200

####
#### Exercise:
#### - parallelize the loop so that each process
####   tries different candidate numbers
#### - if a process finds a factor, print it to the screen
####

####
#### set the variables nprocs / procno
####
##solution
nprocs = comm.Get_size()
procno = comm.Get_rank()
##solved

####
#### divide the work over processes
####
##solution
mylow = procno*(maxfactor//nprocs)
myhigh = (procno+1)*(maxfactor//nprocs)
if mylow<2:
    mylow = 3
if procno==nprocs-1:
    myhigh  = maxfactor
##solved

for myfactor in range(
        ##solution
        mylow,myhigh
        ##solved
        ):
    if bignum%myfactor==0:
        print("Processor %d found factor %d" % (procno,myfactor))



